kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java

```diff
@@ -11,134 +11,124 @@
-    private volatile boolean m_enabled;
-    private volatile boolean m_watchdogToStop = false;
+    private static final Logger logger = LoggerFactory.getLogger(WatchdogServiceImpl.class);
+    private static final String FAILED_COMMAND = "Failed to execute command {} {}";
+
+    private static final long THREAD_TERMINATION_TOUT = 1000;
+    private static final String PROPERTY_ENABLED = "enabled";
+    private static final String PROPERTY_PING_INTERVAL = "pingInterval";
+    private static final String PROPERTY_WD_DEVICE = "watchdogDevice";
+    private static final String PROPERTY_WD_ACTION = "watchdogAction";
+    private static final String WATCHDOG_ACTION_SYSTEM_REBOOT = "system.reboot";
+    private static final String WATCHDOG_ACTION_KURA_RESTART = "kura.restart";
+    private static final int DEFAULT_WD_REFRESH = 10000;
+
+    private volatile boolean isEnabled;
+    private volatile boolean watchdogToStop;
+    private int pingInterval = 2000;
     private String watchdogDevice = "/dev/watchdog";
+    private String watchdogAction = WATCHDOG_ACTION_SYSTEM_REBOOT;
+    private volatile boolean configEnabled = false;   // initialized in properties, if false -> no watchdog
+    private List<CriticalComponentImpl> criticalServiceList;
+    private ComponentContext componentContext;
 
-    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
-        this.m_properties = properties;
-        if (properties == null) {
-            s_logger.debug("activating WatchdogService with null props");
-        } else {
-            if (this.m_properties.get("enabled") != null) {
-                this.m_configEnabled = (Boolean) this.m_properties.get("enabled");
-                if (this.m_configEnabled) {
-                    s_logger.debug("activating WatchdogService with watchdog enabled");
-                } else {
-                    s_logger.debug("activating WatchdogService with watchdog disabled");
-                }
-            }
-            if (this.m_properties.get("pingInterval") != null) {
-                this.pingInterval = (Integer) this.m_properties.get("pingInterval");
-            }
-            if (this.m_properties.get("watchdogDevice") != null
-                    && !((String) this.m_properties.get("watchdogDevice")).isEmpty()) {
-                this.watchdogDevice = (String) this.m_properties.get("watchdogDevice");
-            }
-        }
-        s_criticalServiceList = new ArrayList<CriticalComponentImpl>();
-        this.m_enabled = false;
+    private ScheduledFuture<?> pollThreadTask;
+    private ScheduledExecutorService pollThreadExecutor;
 
-        this.m_pollThreadExecutor = Executors.newSingleThreadScheduledExecutor();
+    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
+        criticalServiceList = Collections.synchronizedList(new ArrayList<CriticalComponentImpl>());
+        this.isEnabled = false;
```

Comment:
it's better to set initial values in a single place. If you don't want to do that in a constructor, set them here. This includes removing the initialization of private fields here